File: src/workflowsuite/kpi/client/time/TimeOffsetCalculator.java
package workflowsuite.kpi.client.time;

import java.time.Duration;
import java.time.Instant;

public final class TimeOffsetCalculator implements ITimeOffsetCalculator {


    /**
     * {@inheritDoc}
     */
    @Override
    public TimeSyncData calculateTimeOffset(NtpData data) {
        Instant t1 = data.getRequestTransmission();
        Instant t2 = data.getRequestReception();
        Instant t3 = data.getResponseTransmission();
        Instant t4 = data.getResponseReception();

        Duration d41 = Duration.between(t1, t4);
        Duration d32 = Duration.between(t2, t3);
        long halfDelta = (d41.minus(d32)).toMillis() / 2;
        long offset = Duration.between(t4, t3).toMillis() + halfDelta;

        return new TimeSyncData(Duration.ofMillis(offset), Duration.ofMillis(halfDelta));
    }
}

File: src/workflowSuite/kpiClient/time/TimeOffsetCalculator.kt
package workflowSuite.kpiClient.time

import java.time.Duration

class TimeOffsetCalculator : ITimeOffsetCalculator {

    override fun calculateTimeOffset(data: NtpData): TimeSyncData {
        val t1 = data.requestTransmission
        val t2 = data.requestReception
        val t3 = data.responseTransmission
        val t4 = data.responseReception

        val d41 = Duration.between(t1, t4)
        val d32 = Duration.between(t2, t3)
        val halfDelta = (d41 - d32).toMillis() / 2
        val offset = Duration.between(t4, t3).toMillis() + halfDelta

        return TimeSyncData(Duration.ofMillis(offset), Duration.ofMillis(halfDelta))
    }
}